packages/ra-core/src/controller/useOptimisticQuery.ts

```diff
@@ -0,0 +1,73 @@
+import { useEffect } from 'react';
+// @ts-ignore
+import { useSelector, useDispatch } from 'react-redux';
+
+import { ReduxState } from '../types';
+
+export interface AsyncQueryResponse {
+    data?: any;
+    loading: boolean;
+    loaded: boolean;
+    error?: any;
+}
+
+export interface FetchAction {
+    type: string;
+    payload: any;
+    meta: {
+        fetch: string;
+        resource: string;
+        onSuccess?: any;
+        onFailure?: any;
+    };
+}
+
+/**
+ * @typedef AsyncQueryResponse
+ * @type {Object}
+ * @property {Object} data
+ * @property {boolean} loading
+ * @property {boolean} loaded
+ * @property {Object} error
+ */
+
+/**
+ * Uses an action to call the data provider
+ * and optimistically loads the result from the store.
+ *
+ * @example
+ *
+ * const { data, loading, loaded, error } = useOptimisticQuery(
+ *   { type 'RA/GET_ONE', payload: { id: 123 }, meta: { fetch: GET_ONE, resource: 'posts' } }
+ *   state => state.admin.resources.posts[123]
+ * );
+ *
+ * @param {object} action The action to dispatch, e.g. { type 'RA/GET_ONE', payload: { id: 123 }, meta: { fetch: GET_ONE, resource: 'posts' } }
+ * @param {function} selector The selector function to get the result from the store
+ * @param {number} version The version number. Increase to force a new fetch
+ *
+ * @returns {GetOneResponse} An object with the shape { data, error, loading, loaded }
+ */
+const useOptimisticQuery = (
+    action: FetchAction,
```

Comment:
I'm not so sure about the signature of this one. Should it accept:
- an action returned by an action creator, as I did here, or
- a quadruplet `{ type, resource, payload, meta }` , like the `useQuery` action

In the latter case, we don't need the crud action creators anymore.